### PR TITLE
[upgrades] Abstraction over system packages

### PR DESCRIPTION
**Pull request by amnn**
## Description

Add an abstraction (trait and macro) over system packages.  Now there are a number of methods that need to be added for each one, to get:

- Framework ID
- module BCS bytes
- CompiledModule
- MovePackage
- Object

To avoid us forgetting one, and to simplify adding new ones in future.

## Test Plan

```
$ cargo simtest
$ env SUI_SKIP_SIMTESTS=1 cargo nextest run
```

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -2852,17 +2851,17 @@ impl AuthorityState {
...
+            MoveStdlib::as_modules(),
+            MoveStdlib::transitive_dependencies(),
         ).await else {
             return vec![];
         };
 
         let Some(sui_framework) = self.compare_system_package(
-            SUI_FRAMEWORK_OBJECT_ID,
+            SuiFramework::ID,
             sui_framework_injection::get_modules(self.name),
```
Comment (emmazzz) [inline crates/sui-core/src/authority.rs:2863]: Clarification q: we do the `sui_framework_injection` trick here for sui framework but not move stdlib purely for testing purposes right? When not in simtests, this just returns the actual framework code in the `sui-framework` folder.

Comment (amnn) [reply crates/sui-core/src/authority.rs:2863]: That's correct!

File: crates/sui-framework/src/lib.rs [hunk 2]
```diff
@@ -12,115 +12,93 @@ use sui_types::{
...
-
-    serialized_modules
-        .into_iter()
-        .map(|module| CompiledModule::deserialize(&module).unwrap())
-        .collect()
-});
-
-pub fn get_sui_framework() -> Vec<CompiledModule> {
-    Lazy::force(&SUI_FRAMEWORK).to_owned()
+macro_rules! system_package {
```
Comment (tzakian) [inline crates/sui-framework/src/lib.rs:20]: Minor naming suggestion: `declare_system_package` (or `define_system_package`) might fit a bit better here?

File: crates/sui-framework/src/lib.rs [hunk 3]
```diff
@@ -12,115 +12,93 @@ use sui_types::{
...
+    ($address:expr, $Package:ident, $path:literal, [$($Dep:ident),* $(,)?]) => {
+        pub struct $Package;
+
+        impl SystemPackage for $Package {
+            const ID: ObjectID = ObjectID::from_address($address);
+            const BCS_BYTES: &'static [u8] = include_bytes!(concat!(env!("OUT_DIR"), "/", $path));
```
Comment (tzakian) [inline crates/sui-framework/src/lib.rs:26]: Just wanted to call out that this is assuming the correct version of the system package has already been written out to this location at compile time.

**Review by tzakian — APPROVED** [review]
So much cleaner! Really love this change. 

Left one small naming suggestion (but what's here already is also fine).

File: crates/sui-framework/src/lib.rs (hunk 3 above)
Comment (amnn) [reply crates/sui-framework/src/lib.rs:26]: Yes, this is already handled by the `sui-framework-build` crate (this is pre-existing).

**Pull request merged**